File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if not inorder: return TreeNode(inorder[0])

        inorder_map = { val: idx for idx, val in enumerate(inorder) }

        def help(start: int, end: int) -> Optional[TreeNode]:
            if start > end: return None

            root_val = postorder.pop()
            root = TreeNode(root_val)

            idx = inorder_map[root_val]

            root.right = help(idx+1, end)
            root.left = help(start, idx-1)

            return root
        
        return help(0, len(inorder)-1)

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        size_t n = inorder.size();

        unordered_map<int,int> inorder_map;
        for(int i=0; i<n; i++) inorder_map[inorder[i]] = i;

        function<TreeNode*(int,int)> help = [&](int start, int end) -> TreeNode* {
            if(start > end) return nullptr;

            const int root_val = postorder.back(); postorder.pop_back();
            TreeNode* root = new TreeNode(root_val);

            const int idx = inorder_map[root_val];

            root->right = help(idx+1, end);
            root->left = help(start, idx-1);

            return root;
        };

        return help(0, n-1);
    }
};